New KEP for SCTP support feature

janosi opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
- If this is your first contribution, read our Getting Started guide https://github.com/kubernetes/community#your-first-contribution
- If you are editing SIG information, please follow these instructions: https://git.k8s.io/community/generator
  You will need to follow these steps:
  1. Edit sigs.yaml with your change 
  2. Generate docs with `make generate`. To build docs for one sig, run `make WHAT=sig-apps generate`
-->

KEP document for SCTP support.
Relevant PR: #64973

janosi commented:
/assign @thockin

danwinship commented on keps/sig-network/0015-20180614-SCTP-support.md:
````diff
@@ -0,0 +1,291 @@
+    port: 80
+    targetPort: 9376
+```
+#### NetworkPolicy with SCTP
````
I think SCTP support should be pretty trivial for plugins that implement policy with iptables, OVS, or eBPF, because all the hard parts would be handled by the kernel. Like, for OpenShift, I think the complete diff would be:

    -    } else if *port.Protocol == kapi.ProtocolTCP || *port.Protocol == kapi.ProtocolUDP {
    +    } else if *port.Protocol == kapi.ProtocolTCP || *port.Protocol == kapi.ProtocolUDP || *port.Protocol == kapi.ProtocolSCTP {
             protocol = strings.ToLower(string(*port.Protocol))

danwinship commented on keps/sig-network/0015-20180614-SCTP-support.md:
```diff
@@ -0,0 +1,291 @@
+A userspace SCTP stack implementation cannot work together with the SCTP kernel module (lksctp) on the same node. That is, the loading of the SCTP kernel module must be avoided on nodes where such applications that use userspace SCTP stack are planned to be run. The problem comes with the introduction of the SCTP protocol option for Services with Virtual IP: once such a service is created the rel ...
+
+1. "-p sctp" is not used in the iptables rules, the processing of requests to the Virtual IP is executed purely based on the destination IP address. In case of ipvs the protocol is a mandatory parameter, so ipvs with SCTP rules cannot be used on the node where userspace SCTP applications should run.
+2. Fall back to the user space proxy on those specific nodes. The user space proxy shall also use a user space SCTP stack, of course. Also the iptables rules that direct the client traffic to the userspace proxy must be created without the "-p sctp" option.
```
"of course"? Ah... is it not possible to implement a generic SCTP proxy via the sockets API?

If the userspace proxy needs to use the userspace SCTP stack then that means the "tainting" runs in both directions: if you want to run the userspace proxy on a node for whatever reason, then you can't run containers that expect to use kernel SCTP on that node.

danwinship replied:
And of course if you don't use kernel SCTP support then that kind of kills my earlier claim that NetworkPolicy would be easy to implement because the kernel does all the hard stuff. Pluging with iptables-, OVS-, or eBPF-based NetworkPolicy presumably would not be able to do SCTP at all on nodes without kernel SCTP.

janosi replied:
@thockin Indeed, a user space proxy does not sound good, but on a node where a userspace SCTP app runs we cannot use the SCTP kernel module anyway. So, if we want to have a proxy on those nodes, then we have to implement the SCTP stack for that proxy in the userspace.
@danwinship Exactly, you see it right. On a node where a userspace SCTP application runs we cannot run such applications that would require the SCTP kernel module. And it is true on the other way around, too.

janosi replied:
Thinking about this part again: it will not work. In order to route a request to e.g. a userspace proxy the port on which the userspace proxy listens must be configured in iptables/ipvs. And it means that the protocol must be configured, that loads the kernel module.
Currently I think, there is no nice way to solve the interworking of userspace SCTP stacks and kernel module based services in the same k8s cluster.

m1093782566 commented on keps/sig-network/0015-20180614-SCTP-support.md:
```diff
@@ -0,0 +1,291 @@
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they releate.
+--->
+#### SCTP in Services
```
IPVS should have SCTP support.

m1093782566 commented on keps/sig-network/0015-20180614-SCTP-support.md:
```diff
@@ -0,0 +1,291 @@
+<!---
+This is where we get down to the nitty gritty of what the proposal actually is.
+--->
+### User Stories [optional]
```
I am curious to learn the user stories of having SCTP support?

janosi replied:
Do you mean, why SCTP support is requested at all?

janosi commented on keps/sig-network/0015-20180614-SCTP-support.md:
```diff
@@ -0,0 +1,291 @@
+  - "@janosi"
+owning-sig: sig-network
+participating-sigs:
+  - sig-cloud-provider
```
OK, so it is not enough then that sig-network is the owning-sig.

janosi commented on keps/sig-network/0015-20180614-SCTP-support.md:
```diff
@@ -0,0 +1,291 @@
+  - TBD
+editor: TBD
+creation-date: 2018-06-14
+last-updated: yyyy-mm-dd
```
Do you mean "creation-date" should be the date when I pushed the doc to github first time?

janosi commented on keps/sig-network/0015-20180614-SCTP-support.md:
```diff
@@ -0,0 +1,291 @@
+
+### Non-Goals
+
+It is not a goal here to add SCTP support to load balancers that are provided by cloud providers. I.e. the Kubernetes user can define Services with type=LoadBalancer and Protocol=SCTP, but if the actual load balancer implementation does not support SCTP then the creation of the Service/load balancer fails.
```
As I know, OpenStack supports SCTP as LB protocol. Though, of course we can start with the approach, that for type=LoadBalancer Protocol=SCTP is not allowed, and whenever there is a need a new PR can enable that for the relevant cloud provider.

I will check the NodePort use case.

johnbelamaric commented on keps/sig-network/0015-20180614-SCTP-support.md:
```diff
@@ -0,0 +1,291 @@
+#### SCTP in NetworkPolicy
+The Kubernetes API modification for the NetworkPolicy is obvious.
+In order to utilize the new protocol value the network controller must support it. 
+
```
Yes, we would at least need to use _stcp for SRV. You should probably explicitly state the expectations around the multi-homing functionality. I assume that will *not* be supported? If it were supported, I think we would want something special around the service discovery to differentiate between IPs that are part of the same multi-homed endpoint vs. part of a separate sctp peer. My knowledge of sctp is pretty limited but reading [RFC 4960 section 5.1.2](https://tools.ietf.org/html/rfc4960#section-5.1.2) it looks like an individual name should resolve to a set of addresses that can be used for multi-homing. This is really different from what we do with SRV (an multiple A records) today, which resolve to separate endpoints.

MaximProshin replied:
To implement SCTP multi-homing in K8s, first of all NAT should support SCTP. There is an IETF draft about it: https://datatracker.ietf.org/doc/draft-ietf-tsvwg-natsupp/

janosi commented on keps/sig-network/0015-20180614-SCTP-support.md:
```diff
@@ -0,0 +1,291 @@
+In any case we shall be able to dedicate these nodes for those userspace SCTP applications, or at least, we must achieve that "regular" SCTP user applications are not deployed on these nodes. The solution proposal for this node separation:
+
+- there shall be a new kube-proxy parameter. If the parameter is set, the kube-proxy switches to this new mode of operation (described above) for SCTP services
+- if the new kube-proxy parameter is set the node must be tainted with a new taint, so the scheduler places only such SCTP applications on this node that use userspace SCTP stack. We must avoid the deployment of "regular" SCTP users on this node.
```
Yes, but what we have to achieve is, that the proxy obeys those taints/labels, so the proxy does not execute any workflow which would load the SCTP kernel module on the node.
Another option is, that the whole SCTP support (as it is now in the code) is put behind a cluster level parameter - if SCTP support is allowed, the cluster uses the kernel modules everywhere. If not allowed, the cluster works as without this enhancement, and userspace SCTP stacks can work as they were before.

danwinship commented:
@janosi So do you (or anyone else here) actually want to do SCTP over the "normal" k8s network? (What the multi-network group is calling the "Kubernetes Cluster-Wide Default Network"). Or are you only interested in SCTP over physical NICs / SR-IOV / DPDK / etc? From what I've been able to gather, the telecom use cases for SCTP also have bandwidth/latency constraints such that you wouldn't want the traffic going over any sort of SDN. Is that accurate? (Or if that's not accurate in general, is it at least accurate for the use cases that require userspace SCTP?)

janosi commented on keps/sig-network/0015-20180614-SCTP-support.md:
```diff
@@ -0,0 +1,291 @@
+#### Interworking with applications that use a user space SCTP stack
+A userspace SCTP stack implementation cannot work together with the SCTP kernel module (lksctp) on the same node. That is, the loading of the SCTP kernel module must be avoided on nodes where such applications that use userspace SCTP stack are planned to be run. The problem comes with the introduction of the SCTP protocol option for Services with Virtual IP: once such a service is created the rel ...
+
+1. "-p sctp" is not used in the iptables rules, the processing of requests to the Virtual IP is executed purely based on the destination IP address. In case of ipvs the protocol is a mandatory parameter, so ipvs with SCTP rules cannot be used on the node where userspace SCTP applications should run.
```
There are the following idea alternatives, that may work with both iptables and ipvs:
1. Configure simple IP forwarding for the ClusterIP on host level: ClusterIP -> localhost. The userspace proxy would bind on localhost + the service specific port.
2. Every ClusterIP is configured to the loopback interface of the host. The userspace proxy binds on the ClusterIP+the service specific port.

janosi commented:
@danwinship Our original purpose was to have SCTP based endpoints/pods in the kube-dns. That was the reason why we initiated the relevant k8s issue. 
From the protocol's usage perspective: we already have SCTP usage in our k8s clusters, in production.  We use SCTP on top of "normal" k8s networks, though we have the capability to provide multiple networks per pod. We do not utilise the accelerator techniques you listed in every product. Some products may need such accelerators, some others not. Also it is up to the deployment environment, capacity requirements, maintaining the network stack, etc. 
Though with the current capabilities of k8s we have to maintain our own service discovery methods for those endpoints/pods, and we would like to simplify this situation.
I am not sure, that high bandwidth and SDN would be mutually exclusive. 
Please note, the "userspace SCTP" mentioned here, and the DPDK (and alike) based user space network stack are not the same. The "userspace SCTP" mentioned here is a raw socket or UDP-encapsulation based solution (usually, as I understand), but the underlying network stack is the regular kernel network stack. Or of course it could be anything (DPDK...), as the "userspace SCTP" stack just sits on top of any IP stack that provides the usual Linux/BSD API (and syscalls) to manage the underlying protocol layers. The "userspace SCTP" requirement came from @MaximProshin in the PR, i.e. it is not our "native" use case in the original issue, but because k8s is a community project, I wanted to have that requirement listed here, so it is not ignored or something (I do not want to polish myself, just explaining the situation, so everyone understands the source of the requirements).

danwinship commented:
> Please note, the "userspace SCTP" mentioned here, and the DPDK (and alike) based user space network stack are not the same.

Yeah, I know. My questions were trying to figure out if the userspace SCTP use case needs to involve "kubernetes networking", or if it just involves "kubernetes" and "networking" but not "kubernetes networking" (and DPDK was an example of something that isn't "kubernetes networking").

I think the userspace SCTP case creates a lot of problems, and if you don't personally care about it, you might want to just drop that stuff from this proposal, and only worry about (a) not implementing things in a way that would make it harder to add userspace SCTP support later, and (b) not implementing things in a way that would break people who were previously successfully using userspace SCTP behind kubernetes's back.

Eg, for the latter, we could make sure that kubernetes itself does not create any `IPPROTO_SCTP` sockets or `-p sctp` iptables rules unless it sees a Pod or Service with an SCTP port; and say that plugins should likewise try to avoid causing the kernel sctp module to be loaded if they don't actually need it.

bowei commented on keps/sig-network/0015-20180614-SCTP-support.md:
```diff
@@ -0,0 +1,192 @@
+
+#### Headless Service with SCTP
+As a user of Kubernetes I want to define headless Services for my applications that use SCTP as L4 protocol on their interfaces, so client applications can discover my applications in kube-dns, or via any other service discovery method that gets information about endpoints via the Kubernetes API.
+Example:
```
can you give the example records that would be added?

janosi replied:
I added an example now. Please check and comment whether that fits for you.

bowei commented on keps/sig-network/0015-20180614-SCTP-support.md:
````diff
@@ -0,0 +1,291 @@
+      port: 7777
+```
+#### User space SCTP stack
+As a user of Kubernetes I want to deploy and run my applications that use a user space SCTP stack. 
````
It feels like this could be separated out from the KEP which is mostly k8s API focused, unless there is some sort of interaction...

bowei commented on keps/sig-network/0015-20180614-SCTP-support.md:
```diff
@@ -0,0 +1,192 @@
+The Kubernetes API modification for the NetworkPolicy is obvious.
+In order to utilize the new protocol value the network controller must support it.
+
+#### Interworking with applications that use a user space SCTP stack
```
This feels like a separate KEP ? to enable use of user-space protocol stacks + kube-proxy.

janosi commented:
@bowei @MaximProshin I added alternative proposals to solve the userspace SCTP stack interworking.  Also I added the explanation why the userspace and kernel space SCTP stack cannot work together currently. Please check those.

KomorkinMikhail commented:
@janosi Thanks for analysing the interworing issue and pointing out the above alternatives.
I suppose that the easiest way for introducing this kind of "userspace proxy" on dedicated nodes is to reuse the same iptables creation logic but without socket creation/listening. This way dedicated nodes will have needed ip rules for user space SCTP service but without LKSCTP. 
It may be some kube-proxy startup option that will be switched on on dedicated nodes. And then checked during service creation:
- if option is turned off: add iptables/ipvs rules, load up kernel sctp, listen to port (LKSCTP nodes)
- if option is turned on: just add iptables/ipvs rules (user space nodes)

Another way is as I mentioned in https://github.com/kubernetes/kubernetes/pull/64973#issuecomment-402065558 introduce an option for service itself alowing user to decide if creation of sctp socket is needed.

janosi commented:
@KomorkinMikhail Thank yo for checking the document!
Unfortunately, the creation of the iptables/ipvs rules triggers the loading of lksctp, too. So, we must avoid the creation of those. I.e. it is not enough to avoid the creation of an SCTP socket.

idealhack commented:
/ok-to-test

KomorkinMikhail commented:
@janosi Great!
I've also tried to figure out can we somehow ignore LKSCTP on nodes containing pods with usercpace SCTP, for example drop LKSCTP abort packets with additional iptables rules. But looks like there's no way we can identify whether it's an LKSCTP packet or ours anyway (even the sending IP is the same as the container IP).

janosi commented:
@KomorkinMikhail Yes, in some cases an ABORT is necessary, and on iptables level we cannot really tell why the ABORT was initiated.
I wonder if the kernel module could check that the packet was also delivered to a raw socket and it would not send ABORT in that case, but it would just ignore ("silently discard") that. 
However, I think, if we can use the same logic here which was used previously without k8s (dedication of nodes) then we are fine.

thockin commented:
I dislike it but if that is the only thing preventing it from working, I could probably forego it.

But we should only be opening a port if teh Service is type=NodePort or LoadBalancer, which we already said we won't do, right?

janosi commented:
@thockin In case of the kube-proxy the openLocalPort() is called also in the case when externalIPs is specified in the Service, and the externalIP matches a local IP of the node.

On the type=NodePort there was indeed the recommendation that we should handle that like we do the type=LoadBalancer case. Though I do not fully understand the reason to exclude the type=NodePort Services. I mean, in the case of type=LoadBalancer it is clear, the cloud providers do not support SCTP as protocol in their load balancer configs. But what is the blocking factor for NodePort?
If we exclude NodePort, should we also exclude the usage of externalIPs with SCTP?

On the other hand if we exclude all these we have the following Service types supported with SCTP:
- headless service without externalIP
- ClusterIP

I wonder then if anyone is eager to have ClusterIP support for SCTP....if not, we get back to the starting point: SCTP support for headless services. We do not need then new kube-proxy parameters. And I think then we can also eliminate the support of SCTP in the pod specification at the containers section (port mapping for containers). We do not hurt then userspace SCTP applications, as all the code that triggers the lksctp loading can be removed.

thockin commented:
I don't know SCTP's vagaries.  Is it "safe" to load-balance SCTP the same
way we do TCP and UDP?  Do we have a real use-case for either NodePort or
External IPs?



On Fri, Jul 6, 2018 at 2:44 PM Laszlo Janosi <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> In case of the kube-proxy the
> openLocalPort() is called also in the case when externalIPs is specified in
> the Service, and the externalIP matches a local IP of the node.
>
> On the type=NodePort there was indeed the recommendation that we should
> handle that like we do the type=LoadBalancer case. Though I do not fully
> understand the reason to exclude the type=NodePort Services. I mean, in the
> case of type=LoadBalancer it is clear, the cloud providers do not support
> SCTP as protocol in their load balancer configs. But what is the blocking
> factor for NodePort?
> If we exclude NodePort, should we also exclude the usage of externalIPs
> with SCTP?
>
> On the other hand if we exclude all these we have the following Service
> types supported with SCTP:
>
>    - headless service without externalIP
>    - ClusterIP
>
> I wonder then if anyone is eager to have ClusterIP support for SCTP....if
> not, we get back to the starting point: SCTP support for headless services.
> We do not need then new kube-proxy parameters. And I think then we can also
> eliminate the support of SCTP in the pod specification at the containers
> section (port mapping for containers). We do not hurt then userspace SCTP
> applications, as all the code that triggers the lksctp loading can be
> removed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/2276#issuecomment-403154190>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVHIdPJSp4j_iVnatDZGK5opgzB_Bks5uD9ofgaJpZM4UqkhH>
> .
>

KomorkinMikhail commented:
@thockin @janosi
As I said we have  implemeted the same SCTP support locally but without openLocalPort(). We have been experimenting with NodePort type of service, we also tried to set the ExternalIP for our service (since we need our service to be available from the external network). Loadbalancing works fine, the same way as in TCP/UDP case.
The only additional thing needed for correct functioning are two conntrack kernel modules: nf_conntrack_proto_sctp and nf_nat_proto_sctp. This way we make sure that all SCTP traffic within one connection is served by one exact POD.

janosi commented:
@thockin @KomorkinMikhail 
This would be my proposal then:

- we do not allow the value "SCTP" for Services with type=LoadBalancer. K8s API validator rejects such services
- we allow other types of Services with value "SCTP"
- we do not allow the value "SCTP" in the pod/container definition as "hostPort" value
- we allow the value "SCTP" in the pod/container definition as "containerPort" value

KomorkinMikhail commented:
@janosi
I haven't looked into LoadBalancer services but I've checked the statement about it in your analysis. Isn't there any requirements on handling of cases of unsupported protocols for cloud providers? At least it seems like a pretty obvious check. I've also checked with my colleague, he approves that all supported cloud providers support such handling.
I am planning to check the code of some lbs

janosi commented:
@KomorkinMikhail Indeed, and you can find my analysis in the PR. LB providers more or less covered the requirement, that they should filter protocols before the request go to the actual cloud's LB controller. Though, as I understood, @thockin asked for blocking the Service creation with type=LB at validation.

janosi commented:
@KomorkinMikhail I think we have to turn to @thockin for the answer. This was his suggestion: "Do we know of any that work? We should probably default to a validation error if this combination is specified."

KomorkinMikhail commented:
@janosi @thockin 
I beleive there's no need in internal blocking. I've checked the code of supported loadbalancers and they do handle cases of unsupported protocols:
`// The service controller verified all the protocols   match on the ports, just check and use the first one
--
if ports[0].Protocol != v1.ProtocolTCP && ports[0].Protocol != v1.ProtocolUDP {
return "", fmt.Errorf("Invalid protocol %s, only TCP and UDP are supported", string(ports[0].Protocol))`

K8s can be used in many other Cloud providers in addition to supported ones. And we are also looking into posibility of some loadbalancer options in future.

janosi commented:
@KomorkinMikhail @thockin I am fine with any of those options. 
Though, for example there was no such protocol check for the GCE internal load balancer, so the rejection would came from the GCE API side in that case.
What I need is an agreement, so I can implement the selected way. 
Also k8s is an open source SW. I believe, that the current SCTP implementation will be modified in the future. My understanding is, that the blocking based solution brings lower risk for the first introduction of SCTP support. Then it can be enhanced with time if needed.

KomorkinMikhail commented:
@janosi Good, let's wait for response from @thockin

janosi commented:
Can anyone please suggest how to move this KEP forward? Will there be a sig-network meeting next week?
@kubernetes/sig-network-pr-reviews @kubernetes/sig-network-proposals

thockin commented on keps/sig-network/0015-20180614-SCTP-support.md:
```diff
@@ -0,0 +1,258 @@
+
+It is not a goal to support SCTP as protocol value for the container's HostPort. The reason: [the usage of HostPort is not recommended by Kubernetes][], and to ensure proper interworking of HostPort with userspace SCTP stacks (see below) would require an additional kubelet/kubenet configuration option. In order to keep the complexity and impact of the introduction of SCTP on a lower level we do n ...
+
+[the usage of HostPort is not recommended by Kubernetes]:https://kubernetes.io/docs/concepts/configuration/overview/#services
```
I don't see why not, I guess.

janosi replied:
OK. The reason I proposed this way was, that listening on HostPorts (in order to reserve the port) is not handled by the kube-proxy but by the kubelet (kubelet -> Docker Service -> Kubenet plugin -> hostportmanager). That is, in order to have the same "userspace sctp config" based listening on HostPorts it would require another config parameter for another component - i.e. the user shall configure the userspace sctp support in 2 different places. And because I understood the cited k8s documentation so, that the usage of HostPorts is not recommended, I thought that then I can skip the support of SCTP for HostPorts. But of course, if the decision is to have SCTP support for HostPorts, too, then I do it.

thockin replied:
In truth, I am not sure how valuable kubelet's logic is here.  ISTR there was/is a bug in it that it doesn't re-open ports if Kubelet restarts, and nobody has complained about it.

Maybe we should just remove that logic altogether?  We could start by skipping it for SCTP and have a separate discussion about removal...  Just make sure it is well commented why we do it for everything except SCTP.

janosi replied:
Fine for me.

thockin commented on keps/sig-network/0015-20180614-SCTP-support.md:
```diff
@@ -0,0 +1,258 @@
+
+### Non-Goals
+
+It is not a goal here to add SCTP support to load balancers that are provided by cloud providers. 
```
I guess I can hold my nose on this.  I really dislike things that are quietly non-portable, like this, but it's hardly the first or worst.

danwinship replied:
Arguably my position on SCTP LoadBalancers here contradicts my position on NetworkPolicy ipBlock ingress on the list... (Although I feel like ipBlock ingress is more broken than SCTP LoadBalancing is so it's not *too* inconsistent.) Having stronger guidelines for situations like this would be good... I think sig-network outsources more behavior to plugins and third parties than most other SIGs do so this is probably something that affects us more than everyone else.

thockin commented on keps/sig-network/0015-20180614-SCTP-support.md:
```diff
@@ -0,0 +1,258 @@
+
+As we can easily see, it is pretty easy to separate application pods that use a userspace SCTP stack from those application pods that use the kernel space SCTP stack: the usual nodeselector label based mechanism,  or taints are there for this very purpose. 
+
+The real challenge here is to ensure that when an SCTP Service is created in a Kubernetes cluster the Kubernetes logic does not create listening SCTP sockets on those nodes that are dedicated for the applications that use  userspace SCTP stack - because such an action would trigger the loading of the kernel module.
```
It's not clear to me.

Is this proposal saying that Kubelet and kube-proxy will not do the usual "open and hold" operation for SCTP ports?

Is that sufficient?  If I have a cluster with an SCTP service, that triggers loading iptables modules for sctp, but does not listen() on the port - will userspace SCTP be OK?

danwinship replied:
> Is this proposal saying that Kubelet and kube-proxy will not do the usual "open and hold" operation for SCTP ports?

Yeah, on nodes that were marked userspace-sctp. Which would mean that HostPort/ExternalIP SCTP services wouldn't work on userspace-sctp nodes, and NodePort SCTP services would not be reachable on userspace-sctp nodes.

But that weirdness would be limited to clusters that used userspace sctp; in clusters with no userspace sctp nodes, SCTP would behave just like everything else.

> If I have a cluster with an SCTP service, that triggers loading iptables modules for sctp, but does not listen() on the port - will userspace SCTP be OK?

It looks like the kernel has some code to parse SCTP packets even without the `sctp` module being loaded, and you can add rules with matches like `-p sctp --dport 80` without needing the kernel sctp module. (It will cause `xt_sctp` to be loaded, but that's just part of iptables, and wouldn't interfere with userspace SCTP processing.)

It is possible that some iptables extension rules (NAT?) might cause `sctp` to be loaded.

thockin replied:
I'm not very excited by formalizing userspace SCTP as a concept in Kubernetes based on labels.  In general, we don't want to use annotations or labels for config, when we have real config (even if kube-proxy is behind the curve wrt component config).

We could add a flag/component-config param here, and leave labels to users.  I would be OK with that.

janosi replied:
Wrt SCTP NAT: according to our observations, iptables based NAT does not cause any harm. As @KomorkinMikhail wrote above in this PR, when they implemented SCTP related patches into k8s they simply skipped those "openlocalport" parts for SCTP, and they could use it with NodePort and ExternalIP as well.

thockin replied:
OK.  This wasn't clear.  The main points, if I understand:

1) skip the opening/holding of the port in both kubelet and kube-proxy IFF SCTP
2) suggest that user segregate their user-space SCTP work and their kernel-space SCTP work by node labels, but the details of that are left to the user

janosi replied:
Yepp, my English.... :/
Exactly, that was the plan.

janosi commented on keps/sig-network/0015-20180614-SCTP-support.md:
```diff
@@ -0,0 +1,258 @@
+      role: myservice
+  policyTypes:
+  - Ingress
+  - Egress
```
OK.

janosi commented on keps/sig-network/0015-20180614-SCTP-support.md:
````diff
@@ -0,0 +1,258 @@
+    - protocol: SCTP
+      port: 7777
+```
+#### Userspace SCTP stack
````
All right. For me the justification came from the fact that there are apps that use userspace SCTP for any reason right now, and most probably they do not want to re-write their network handling logic at the same time when they move to containers+k8s - so they must have userspace SCTP support.

janosi commented:
@thockin @danwinship I updated the document according to the comments above and the discussion in the SIG Network meeting yesterday. Please check the document, and if it is OK, please move it forward into a next state.

m1093782566 commented:
I am interested in joining this effort and can take the service/kube-proxy implementation of SCTP if no one is working on this ...

janosi commented on keps/sig-network/0015-20180614-SCTP-support.md:
````diff
@@ -0,0 +1,292 @@
+  - 80.11.12.10
+```
+
+#### NetworkPolicy with SCTP
````
I do so.

andrewsykim commented on keps/sig-network/0015-20180614-SCTP-support.md:
```diff
@@ -0,0 +1,292 @@
+
+### Non-Goals
+
+It is not a goal here to add SCTP support to load balancers that are provided by cloud providers. The Kubernetes side implementation will not restrict the usage of SCTP as the protocol for the Services with type=LoadBalancer, but we do not implement the support of SCTP into the cloud specific load balancer implementations.
```
Will do! Thanks!
